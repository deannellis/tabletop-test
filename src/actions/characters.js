import uuid from 'uuid';

// ADD_CHARACTER
export const addCharacter = (
    {
        abilities = {
            strength: 0,
            intelligence: 0,
            wisdom: 0,
            dexterity: 0,
            constitution: 0,
            charisma: 0
        }, 
        race = '', 
        charClass = '', 
        hp = 0, 
        xp = 0,
        name = '',
        description = ''
    } = {}
) => ({
    type: 'ADD_CHARACTER',
    character: {
        id: uuid(),
        abilities,
        race,
        charClass,
        hp,
        xp,
        name,
        description
    }
});

// REMOVE_CHARACTER

export const removeCharacter = ({ id } = {}) => ({
    type: 'REMOVE_CHARACTER',
    id
});

// EDIT_CHARACTER
export const editCharacter = (id, updates) => ({
    type: 'EDIT_CHARACTER',
    id,
    updates
});

