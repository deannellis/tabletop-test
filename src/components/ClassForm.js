import React from 'react';
import { connect } from 'react-redux';
import spells from '../objects/spells';
import SpellListItem from './SpellListItem';

class ClassForm extends React.Component {
    constructor(props){
        super(props);
        
        this.state = {
            selectedOption: '',
            error: ''
        };
    }
    
    render() {
        const characters = this.props.characters;
        const lastCharacter = characters[characters.length - 1];
        let canCleric = true;
        let canFighter = true;
        let canMagicUser = true;
        let canThief = true;
        
        if(lastCharacter.race == 'dwarf') {
            canMagicUser = false;
        } else if (lastCharacter.race == 'halfling') {
            canFighter = false;
        } 
        
        if(lastCharacter.abilities.wisdom < 9) {
            canCleric = false;
        }
        
        if(lastCharacter.abilities.strength < 9) {
            canFighter = false;
        }
        
        if(lastCharacter.abilities.intelligence < 9) {
            canMagicUser = false;
        }
        
        if(lastCharacter.abilities.dexterity < 9) {
            canThief = false;
        }
        
        return (
            <div>
                <h2>Select Your Class</h2>
                {this.state.error && <p>{this.state.error}</p>}
                <form onSubmit={this.onSubmit} >
                    {canCleric 
                        ? 
                        <div>
                            <label>
                                <input type="radio" name="class" value="cleric"
                                    checked={this.state.selectedOption === 'cleric'}
                                    onChange={this.handleOptionChange}
                                />
                                Cleric
                            </label>
                        </div> 
                        : 
                        <p>Cleric: Your ability scores are insufficient</p>
                    }
                    {canFighter 
                        ? 
                        <div>
                            <label>
                                <input type="radio" name="class" value="fighter"
                                    checked={this.state.selectedOption === 'fighter'}
                                    onChange={this.handleOptionChange}
                                />
                                Fighter
                            </label>
                        </div> 
                        : 
                        <p>Fighter: Your ability scores are insufficient</p>
                    }
                    {canMagicUser 
                        ? 
                        <div>
                            <label>
                                <input type="radio" name="class" value="magic-user"
                                    checked={this.state.selectedOption === 'magic-user'}
                                    onChange={this.handleOptionChange}
                                />
                                Magic-User
                            </label>
                        </div> 
                        : 
                        <p>Magic-User: Your ability scores are insufficient</p>
                    }
                    {canThief 
                        ? 
                        <div>
                            <label>
                                <input type="radio" name="class" value="thief"
                                    checked={this.state.selectedOption === 'thief'}
                                    onChange={this.handleOptionChange}
                                />
                                Thief
                            </label>
                        </div> 
                        : 
                        <p>Thief: Your ability scores are insufficient</p>
                    }
                    <button>Next Step</button>
                </form>
            </div>
        )
    }
    
    handleOptionChange = (e) => {
        this.setState({
            selectedOption: e.target.value
        });
    }
    
    onSubmit = (e) => {
        e.preventDefault();
        if(!this.state.selectedOption) {
            this.setState(() => ({
                error: 'Please select a class'
            }));
        } else {
            const characters = this.props.characters;
            const lastCharacter = characters[characters.length - 1];
            this.setState(() => ({
                error: ''
            }));
            this.props.onSubmit( lastCharacter.id, {
                charClass: this.state.selectedOption
            });
        }
    }
}

const mapStateToProps = (state) => {
    return {
        characters: state.characters
    };
};

export default connect(mapStateToProps)(ClassForm);