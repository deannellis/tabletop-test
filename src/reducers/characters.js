//  CHARACTERS REDUCER
import actionIds from 'actions/actionIds';
const {ADD_CHARACTER, REMOVE_CHARACTER, EDIT_CHARACTER, COMPLETE_CHARACTER} = actionIds;
const charactersReducerDefaultState = [];

export default ( state = charactersReducerDefaultState, action ) => {
    switch (action.type) {
        case ADD_CHARACTER:
            return [
                ...state,
                action.character
            ];
        case REMOVE_CHARACTER:
            return state.filter(({ id }) => {
                return id !== action.id;
            });
        case EDIT_CHARACTER:
            return state.map((character) => {
                if (character.id ===  action.id) {
                    return {
                        ...character,
                        ...action.updates
                    }
                } else {
                    return character;
                }
            });
        case COMPLETE_CHARACTER:
            return state.map((character) => {
                if (character.id ===  action.id) {
                    let {inProgressStep, ...rest} = character;
                    console.log('rest'+rest);
                    return rest;
                } else {
                    return character;
                }
            });
        default:
            return state;
    }
};