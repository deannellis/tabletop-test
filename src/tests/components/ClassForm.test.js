/* global expect */
import React from 'react';
import { shallow } from 'enzyme';
import ClassForm from '../../components/ClassForm';
import characters from '../fixtures/characters';

test('should render class form without props (character not found)', () => {
    const wrapper = shallow(<ClassForm />);
    expect(wrapper).toMatchSnapshot();
});

test('should render class form with dwarf (magic-user disabled)', () => {
    const wrapper = shallow(<ClassForm currentCharacter={characters[1]}/>);
    expect(wrapper).toMatchSnapshot();
});

test('should render class form with halfling (figther disabled)', () => {
    const wrapper = shallow(<ClassForm currentCharacter={characters[2]}/>);
    expect(wrapper).toMatchSnapshot();
});

test('should render class form with low wisdom (cleric disabled)', () => {
    const wrapper = shallow(<ClassForm currentCharacter={characters[3]}/>);
    expect(wrapper).toMatchSnapshot();
});

test('should render class form with low strength (figther disabled)', () => {
    const wrapper = shallow(<ClassForm currentCharacter={characters[4]}/>);
    expect(wrapper).toMatchSnapshot();
});

test('should render class form with low intelligence (magic-user disabled)', () => {
    const wrapper = shallow(<ClassForm currentCharacter={characters[5]}/>);
    expect(wrapper).toMatchSnapshot();
});

test('should render class form with low dexterity (thief disabled)', () => {
    const wrapper = shallow(<ClassForm currentCharacter={characters[6]}/>);
    expect(wrapper).toMatchSnapshot();
});