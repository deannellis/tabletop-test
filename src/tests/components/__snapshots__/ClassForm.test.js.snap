// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render class form with dwarf (magic-user disabled) 1`] = `
<div>
  <h2>
    Select Your Class
  </h2>
  <form
    onSubmit={[Function]}
  >
    <div>
      <label>
        <input
          checked={false}
          name="class"
          onChange={[Function]}
          type="radio"
          value="cleric"
        />
        Cleric
      </label>
    </div>
    <div>
      <label>
        <input
          checked={false}
          name="class"
          onChange={[Function]}
          type="radio"
          value="fighter"
        />
        Fighter
      </label>
    </div>
    <p>
      Magic-User: Dwarves cannot be Magic-Users
    </p>
    <div>
      <label>
        <input
          checked={false}
          name="class"
          onChange={[Function]}
          type="radio"
          value="thief"
        />
        Thief
      </label>
    </div>
    <button>
      Next Step
    </button>
  </form>
</div>
`;

exports[`should render class form with halfling (figther disabled) 1`] = `
<div>
  <h2>
    Select Your Class
  </h2>
  <form
    onSubmit={[Function]}
  >
    <div>
      <label>
        <input
          checked={false}
          name="class"
          onChange={[Function]}
          type="radio"
          value="cleric"
        />
        Cleric
      </label>
    </div>
    <p>
      Fighter: Halflings cannot be fighters
    </p>
    <div>
      <label>
        <input
          checked={false}
          name="class"
          onChange={[Function]}
          type="radio"
          value="magic-user"
        />
        Magic-User
      </label>
    </div>
    <div>
      <label>
        <input
          checked={false}
          name="class"
          onChange={[Function]}
          type="radio"
          value="thief"
        />
        Thief
      </label>
    </div>
    <button>
      Next Step
    </button>
  </form>
</div>
`;

exports[`should render class form with low dexterity (thief disabled) 1`] = `
<div>
  <h2>
    Select Your Class
  </h2>
  <form
    onSubmit={[Function]}
  >
    <div>
      <label>
        <input
          checked={false}
          name="class"
          onChange={[Function]}
          type="radio"
          value="cleric"
        />
        Cleric
      </label>
    </div>
    <div>
      <label>
        <input
          checked={false}
          name="class"
          onChange={[Function]}
          type="radio"
          value="fighter"
        />
        Fighter
      </label>
    </div>
    <div>
      <label>
        <input
          checked={false}
          name="class"
          onChange={[Function]}
          type="radio"
          value="magic-user"
        />
        Magic-User
      </label>
    </div>
    <p>
      Thief: Your dexterity score is too low
    </p>
    <button>
      Next Step
    </button>
  </form>
</div>
`;

exports[`should render class form with low intelligence (magic-user disabled) 1`] = `
<div>
  <h2>
    Select Your Class
  </h2>
  <form
    onSubmit={[Function]}
  >
    <div>
      <label>
        <input
          checked={false}
          name="class"
          onChange={[Function]}
          type="radio"
          value="cleric"
        />
        Cleric
      </label>
    </div>
    <div>
      <label>
        <input
          checked={false}
          name="class"
          onChange={[Function]}
          type="radio"
          value="fighter"
        />
        Fighter
      </label>
    </div>
    <p>
      Magic-User: Your intelligence score is too low
    </p>
    <div>
      <label>
        <input
          checked={false}
          name="class"
          onChange={[Function]}
          type="radio"
          value="thief"
        />
        Thief
      </label>
    </div>
    <button>
      Next Step
    </button>
  </form>
</div>
`;

exports[`should render class form with low strength (figther disabled) 1`] = `
<div>
  <h2>
    Select Your Class
  </h2>
  <form
    onSubmit={[Function]}
  >
    <div>
      <label>
        <input
          checked={false}
          name="class"
          onChange={[Function]}
          type="radio"
          value="cleric"
        />
        Cleric
      </label>
    </div>
    <p>
      Fighter: Your strength score is too low
    </p>
    <div>
      <label>
        <input
          checked={false}
          name="class"
          onChange={[Function]}
          type="radio"
          value="magic-user"
        />
        Magic-User
      </label>
    </div>
    <div>
      <label>
        <input
          checked={false}
          name="class"
          onChange={[Function]}
          type="radio"
          value="thief"
        />
        Thief
      </label>
    </div>
    <button>
      Next Step
    </button>
  </form>
</div>
`;

exports[`should render class form with low wisdom (cleric disabled) 1`] = `
<div>
  <h2>
    Select Your Class
  </h2>
  <form
    onSubmit={[Function]}
  >
    <p>
      Cleric: Your wisdom score is too low
    </p>
    <div>
      <label>
        <input
          checked={false}
          name="class"
          onChange={[Function]}
          type="radio"
          value="fighter"
        />
        Fighter
      </label>
    </div>
    <div>
      <label>
        <input
          checked={false}
          name="class"
          onChange={[Function]}
          type="radio"
          value="magic-user"
        />
        Magic-User
      </label>
    </div>
    <div>
      <label>
        <input
          checked={false}
          name="class"
          onChange={[Function]}
          type="radio"
          value="thief"
        />
        Thief
      </label>
    </div>
    <button>
      Next Step
    </button>
  </form>
</div>
`;

exports[`should render class form without props (character not found) 1`] = `
<div>
  Character not found
  <div>
    <Link
      replace={false}
      to="/"
    >
      Home
    </Link>
  </div>
</div>
`;
